### ISSUE TITLE
Do we need to increase number of epoch when we are using model.fit_generator instead of model.fit?

### ISSUE THREAD

**Issue by meysamgolm** [fchollet/keras#4757]
I am already aware of some discussions on how to use Keras for very large datasets. But I saw something interesting! When I am using model.fit, my loss decreases to 0.3 with number_epoch=4. But when I am using model.fit_generator, my loss decreases to 0.3 by number_epoch=12. Do you have any idea?

I put my script here just in case:

`def generate_reduced_dimension_arrays_from_file(ilist_a, edfroot_a, featroot_a, labroot_a, feat_ext_a, lab_ext_a, num_channels_a, feat_fdur_a, win_len_a, models_order, ipca_model):

    while True:

        # loop over all input list files 
        #
        for i in range(0, len(ilist_a)):

            # read a list of edfs into memory
            #
            edf_list = [line.rstrip('\n') for line in open(ilist_a[i])]
    
            # loop over all the edf files inside the list.
            #
            for edf_path in edf_list:

                # numpy arrays befor applying IPCA.
                #
                d_raw_np = []
                a_raw_np = []
                train_set_x = []
                train_set_y = []
                d_ipca = []
                d_raw_np, a_raw_np = generate_arrays_from_file(edf_path, edfroot_a, featroot_a, labroot_a, feat_ext_a, lab_ext_a, num_channels_a, feat_fdur_a, win_len_a, models_order)

                # apply IPCA on data 
                #
                d_ipca = ipca_model.transform(d_raw_np)
                print "d_ipca.shape = " , d_ipca.shape

                # reshape data and annotation for LSTM.
                # This is a requirement for Keras.
                # 
                train_set_x = np.reshape(d_ipca, d_ipca.shape + (1,))
                train_set_y = np.reshape(a_raw_np, a_raw_np.shape + (1,))
                yield (train_set_x, train_set_y)

...

    # define LSTM model
    #
    print "Build LSTM model ..."
    lstm_model = Sequential()
    lstm_model.add(LSTM(lstm_batch_size_a, input_shape=(ipca_output_size_a, 1)))
    lstm_model.add(Dense(1, activation='sigmoid'))
    lstm_model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    print(lstm_model.summary())
    lstm_model.fit_generator(generate_reduced_dimension_arrays_from_file(ilist_a, edfroot_a, featroot_a, labroot_a, feat_ext_a, lab_ext_a, num_channels_a, feat_fdur_a, win_len_a,    models_order, ipca_model),
        samples_per_epoch=80000, nb_epoch=12)`

**Comment by patyork** [comment]
You should not, no; all things equal, `fit` and `fit_generator` should behave the same.

To determine why not, a couple of questions:

- Please post the `fit()` code you are comparing against
- How many samples, total, are there in your files, if you were to just open and read all of the files? "Epochs" tells us nothing, unless we know how many samples there are per epoch.

**Comment by meysamgolm** [comment]
I have totally 35372 sample. I ran lots of experiment in last few days and the results showed me that fit and fit_generator do not behave the same. 

Additionally this is the fit() code:

`    d_raw_dlin = []
    a_raw_dlin = []
                
    # read label parameter file for deep learning.
    #
    models, models_order = nru.nedc_get_overall_labels_param(labels_param_a)

    # loop over all input list files 
    #
    for i in range(0, len(ilist_a)):
        
        # read a list of edfs into memory
        #
        edf_list = [line.rstrip('\n') for line in open(ilist_a[i])]
    
        # loop over all the edf files inside the list.
        #
        for edf_path in edf_list:
    
            # extract basename and creat the file path for feature and annotation files.
            #
            print "begin to process", edf_path
            basename = os.path.splitext(os.path.basename(edf_path))[0]

            htk_path = edf_path.replace(edfroot_a,featroot_a)
            htk_path = htk_path.replace(os.path.splitext(os.path.basename(edf_path))[1], ("." + feat_ext_a))
            htk_dir = os.path.dirname(htk_path)

            lab_path = edf_path.replace(edfroot_a,labroot_a)
            lab_path = lab_path.replace(os.path.splitext(os.path.basename(edf_path))[1],("." + lab_ext_a))

            # load features data
            #
            feats_data = nru.nedc_get_feats(basename,
                                            htk_dir,
                                            num_channels_a,
                                            feat_ext_a);

            # load overall annotations
            #
            annotations = nru.nedc_get_overall_annotations(lab_path)

            # calculate the frame rate ratio from the actual frame duration
            # and the value implied by HTK
            #
            fr_ratio = HTK_FDUR / feat_fdur_a


            # compose the input for deep learning input. 
            # In fact here, we copmes the input just for
            # part of data that we have annotations.
            #
            for stime in range(len(annotations)):

                # define empty lists for each element of data and annotations.
                #
                d_raw_dlin_element = []
                a_raw_dlin_element = []


                # extract start time, stop time and the label from annotations.
                #
                start_time = annotations[stime][0]
                stop_time = annotations[stime][1]
                label = annotations[stime][2]
               
                # extract the event id for the related label.
                #
                event_order, a_raw_dlin_element = models_order[label]

                # covert start_time and stop_time from HTK scale
                # to second.
                #
                start_time = start_time / (fr_ratio * HTK_TIME_SCALE)
                stop_time = stop_time / (fr_ratio * HTK_TIME_SCALE)

                # compute start index and stop index for reading HTK files
				# based on window length and frame epoch duration
				#
                winh = int(round(win_len_a - (stop_time - start_time)) / 2)
                start_index = int(round((start_time - winh) / feat_fdur_a))
                stop_index = int(round((stop_time + winh) / feat_fdur_a))
                
                # if start index and stop index remains in the range of feats data
                # then read and append the related data and annotations
                #
                if (start_index >= 0 and stop_index < len(feats_data[0])):
                    for tindex in range(start_index, stop_index):                
                        for ch in range(num_channels_a):
                            for ft in range(len(feats_data[ch][tindex])):
                                d_raw_dlin_element.append(feats_data[ch][tindex][ft])

                    # apend the related elements to lists for data and annotations.
                    #
                    d_raw_dlin.append(d_raw_dlin_element)
                    a_raw_dlin.append(a_raw_dlin_element)

    # convert lists of data and annotations to numpy arrays
    #
    d_raw_np = np.asarray(d_raw_dlin)
    a_raw_np = np.asarray(a_raw_dlin)
    print "d_raw_np.shape = " , d_raw_np.shape
    print "a_raw_np.shape = " , a_raw_np.shape


    # apply IPCA on data read from HTK.
    #

    # train IPCA model.
    #
    print "Build IPCA model ..."
    ipca = IncrementalPCA(batch_size=ipca_batch_size_a, n_components=ipca_output_size_a, copy=True)
    d_ipca = ipca.fit(d_raw_np)
    d_ipca = ipca.transform(d_raw_np)
    print "d_ipca.shape = " , d_ipca.shape

    # save IPCA model.
    #
    print "Save IPCA model to " + odir_a + "/ipca_model" + " ..."
    ipca_output_model_file = odir_a + "/ipca_model"
    model_file = open(ipca_output_model_file, "wb") 
    cPickle.dump(ipca, model_file, protocol=cPickle.HIGHEST_PROTOCOL)
    model_file.close()

    # train LSTM deep learning network by the output of IPCA.
    #
    print "Build LSTM model ..."

    # reshape data and annotation for LSTM.
    # This is a requirment for Keras.
    #
    train_set_x = np.reshape(d_ipca, d_ipca.shape + (1,))
    train_set_y = np.reshape(a_raw_np, a_raw_np.shape + (1,))
    print "train_set_x_shape=", train_set_x.shape
    print "train_set_y_shape=", train_set_y.shape

    # build LSTM and train it.
    #
    model = Sequential()
    model.add(LSTM(lstm_batch_size_a, input_shape=train_set_x.shape[1:]))
    model.add(Dense(1, activation='sigmoid'))
    model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
    print(model.summary())
    model.fit(train_set_x, train_set_y, validation_data=(train_set_x, train_set_y),
          nb_epoch=lstm_epoch_size_a, batch_size=lstm_batch_size_a)`
